backend/app/user/supabase/supabase_user_service.py

```diff
@@ -0,0 +1,170 @@
+        self, *, table_name: str, columns: str, eq: dict
+    ) -> APIResponse:
+        """
+        Run a `SELECT` query on a table with the name `table_name`.
+
+        An `equal_to` filter is added to the query via the `eq` parameter.
+        """
+
+        return (
+            self.__database_client.table(table_name).select(columns).eq(**eq).execute()
+        )
+
+    def __upsert_into_database(
+        self, *, table_name: str, json: dict | tuple, constraint: str = ""
+    ) -> APIResponse:
+        """
+        Run an `UPSERT` query on a table with the name `table_name`.
+
+        An `UPDATE` query is run if there is a column conflict with `constraint`.
+
+        An `INSERT` query is run otherwise.
+        """
+
+        upsert_json: dict | list = list(json) if isinstance(json, tuple) else json
+
+        return (
+            self.__database_client.table(table_name)
+            .upsert(upsert_json, on_conflict=constraint)
+            .execute()
+        )
+
+    @override
+    def add_to_user_anti_recommendations_history(
+        self, *, user_id: UserId, anti_recommendation_key: RecordKey
+    ) -> None:
+        """Append `anti_recommendation_key` to a User's anti-recommendation history."""
+
+        try:
+            anti_recommendations_history = list(
+                self.get_user_anti_recommendations_history(user_id)
+            )
+
+            anti_recommendations_history.append(anti_recommendation_key)
+
+            self.__upsert_into_database(
+                table_name=ARKG_ANTI_RECOMMENDER_USER_STATE_TABLE_NAME,
+                json={
+                    "anti_recommendations_history": anti_recommendations_history,
+                },
+                constraint=str(user_id),
+            )
+        except APIError as exception:
+            raise SupabaseUserServiceException(
+                message=f"Unable to add anti-recommendation to history of User with id {user_id}. Encountered database exception: {exception.message}"
+            ) from exception
+
+    @override
+    def get_user_anti_recommendations_history(
+        self, user_id: UserId
+    ) -> tuple[RecordKey, ...]:
+        """Return a tuple containing Record keys in a User's anti-recommendation history."""
+
+        try:
+            database_service_result = self.__fetch_from_database(
+                table_name=ARKG_ANTI_RECOMMENDER_USER_STATE_TABLE_NAME,
+                columns="anti_recommendations_history",
+                eq={"column": "user_id", "value": str(user_id)},
+            )
+        except APIError as exception:
+            raise SupabaseUserServiceException(
+                message=f"Unable to get anti-recommendation history of User with id: {user_id}. Encountered database exception: {exception.message}"
+            ) from exception
+
+        return tuple(database_service_result.data[0]["anti_recommendations_history"])
+
+    @override
+    def get_user_last_seen_anti_recommendation(self, user_id: UserId) -> str:
```

Comment:
`-> RecordKey`

Reply:
Return type for this method has been changed to `RecordKey | None`.